lib/algos/list/delimiterNonCurrent.js

```diff
@@ -131,25 +178,29 @@ class DelimiterNonCurrent extends Delimiter {
         }
         ++this.evaluatedKeys;
 
-        const versionIdIndex = keyVersionSuffix.indexOf(VID_SEP);
-        const key = keyVersionSuffix.slice(0, versionIdIndex);
-        const versionId = keyVersionSuffix.slice(versionIdIndex + 1);
+        const { key, versionId } = this.parseKey(keyVersionSuffix);
 
         this.NextKeyMarker = key;
         this.NextVersionIdMarker = versionId;
 
-        // For a given key, the latest version is skipped since it represents either:
-        // - the current version or
-        // - the last version of the previous truncated listing
-        const isLatestVersion = key !== this.keyName;
+        // The master key serves two purposes:
+        // - It retrieves the expiration date for the previous version that is no longer current.
+        // - It excludes the current version from the list.
+        const isMasterKey = versionId === undefined;
+        if (isMasterKey) {
+            this.masterKey = key;
+            this.masterVersionId = Version.from(value).getVersionId() || 'null';
 
-        if (isLatestVersion) {
-            this.keyName = key;
-            // The current last-modified date is kept in memory and used as a "stale date" for the following version.
             this.staleDate = this.getLastModified(value);
             return FILTER_ACCEPT;
         }
 
+        const isCurrentVersion = this.masterKey === key && this.masterVersionId === versionId;
```

Comment:
If delete markers are also candidates for expiration, they don't have an associated master key even though they can be current. The way to know whether they are current is if the previous key in the merged M+V listing is a different key, or if it's the very first key of the listing.

Reply:
- Delete markers have a version ID that is correctly ordered.
- It doesn't matter if the current delete marker doesn't have a master when listing non-current versions.
The first version key in the listing should store the "delete marker" version and will be skipped.

Reply:
Also, just to make sure, I added tests for delete marker here: https://github.com/scality/cloudserver/pull/5231